Source/FieldSolver/FiniteDifferenceSolver/MacroscopicEvolveHM_2nd.cpp

```diff
@@ -0,0 +1,902 @@
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> &Mfield, // Mfield contains three components MultiFab
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> &Hfield,
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> &Bfield,
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> const &H_biasfield, // H bias
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> const &Efield,
+    amrex::Real const dt,
+    std::unique_ptr<MacroscopicProperties> const &macroscopic_properties) {
+
+    if (m_fdtd_algo == MaxwellSolverAlgo::Yee){
+        MacroscopicEvolveHMCartesian_2nd<CartesianYeeAlgorithm>(Mfield, Hfield, Bfield, H_biasfield, Efield, dt, macroscopic_properties);
+    } else {
+        amrex::Abort("Only yee algorithm is compatible for M updates.");
+    }
+} // closes function MacroscopicEvolveHM_2nd
+#endif
+#ifdef WARPX_MAG_LLG
+template <typename T_Algo>
+void FiniteDifferenceSolver::MacroscopicEvolveHMCartesian_2nd(
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> &Mfield,
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> &Hfield,
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> &Bfield,
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> const &H_biasfield, // H bias
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> const &Efield,
+    amrex::Real const dt,
+    std::unique_ptr<MacroscopicProperties> const &macroscopic_properties) {
+
+    // obtain the maximum relative amount we let M deviate from Ms before aborting
+    amrex::Real mag_normalized_error = macroscopic_properties->getmag_normalized_error();
+
+    auto &warpx = WarpX::GetInstance();
+    int coupling = warpx.mag_LLG_coupling;
+    int M_normalization = warpx.mag_M_normalization;
+
+    // build temporary vector<multifab,3> Mfield_prev, Mfield_error, a_temp, a_temp_static, b_temp_static
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> Hfield_old;    // H^n before the current time step
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> Mfield_old;    // M^n before the current time step
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> Mfield_prev;   // M^(n+1) of the (r-1)th iteration
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> Mfield_error;  // The error of the M field between the twoiterations
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> a_temp;        // right-hand side of vector a, see the documentation
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> a_temp_static; // α M^n/|M| in the right-hand side of vector a, see the documentation
+    std::array<std::unique_ptr<amrex::MultiFab>, 3> b_temp_static; // right-hand side of vector b, see the documentation
+
+    // initialize Mfield_previous
+    for (int i = 0; i < 3; i++){
+        Hfield_old[i].reset(new MultiFab(Hfield[i]->boxArray(), Hfield[i]->DistributionMap(), 3, Hfield[i]->nGrow()));
```

Comment:
```suggestion
        Hfield_old[i].reset(new MultiFab(Hfield[i]->boxArray(), Hfield[i]->DistributionMap(), 1, Hfield[i]->nGrow()));
```

Should this be 3 or 1?